### PR TITLE
Format docstrings as complete chunks

### LINKED ISSUE #240: Darker keep indenting docstring.
Sorry this need to be a full repo for reproduction, 

But basically `darker -r <revision> file` does not converge and keep indenting the docstring on each call

see https://github.com/Carreau/bugd

I'll dive into it when I have time (and yes I still need to look into how to integrate pyupgrade and find more time to contribute to darker)

### PR DESCRIPTION
**Pull request by akaihola**
Treating large docstrings as separate chunks can cause to weird behavior, e.g. "reformat loops" in which some lines are repeatedly indented and unindented by Darker.

Fixes #240.

- [x] add failing test case
- [x] find multi-line using AST parsing, using e.g.:
  ```python
  from pathlib import Path
  from tokenize import tokenize, STRING
  from darker.utils import TextDocument

  src = TextDocument.from_file(Path("setup.py"))
  readline = (f"{line}\n".encode(src.encoding) for line in src.lines).__next__
  multiline_string_linespans = [
      (t.start[0], t.end[0]) for t in tokenize(readline)
      if t.type == STRING and t.end[0] > t.start[0]
  ]
  ```
- [x] if any line in a multi-line string was edited by the user, mark all of the lines edited
- [x] change log
- [x] tests for added functions

### CODE REVIEW

**Comment by akaihola** [comment]
@Carreau, any other ideas for fixing this than doing our own AST parsing to find where possible multi-line docstrings are?

**Comment by akaihola** [comment]
@Carreau, this seems to work now, and I found two edge case bugs in my implementation when completing the unit test coverage! I wonder if a second pair of eyes reviewing this would find even more.

I could include this in version 1.4.2 if you have time to review, or alternatively postpone until the 1.5.0 release.

File: src/darker/tests/test_multiline_strings.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    dict(start=11, end=11, ranges=TEST_RANGES, expect=None),
+    dict(start=11, end=12, ranges=TEST_RANGES, expect=None),
+    dict(start=12, end=12, ranges=TEST_RANGES, expect=None),
+    dict(start=10, end=19, ranges=TEST_RANGES, expect=(9, 17)),
+)
+def test_find_overlap(start, end, ranges, expect):
+    """`find_overlap()` finds the enclosing range for overlapping ranges"""
+    result = multiline_strings.find_overlap(start, end, ranges)
+
+    assert result == expect
```
Comment (MatthijsBurgh) [inline src/darker/tests/test_multiline_strings.py:124]: Maybe, provide a message, to show the values/diff. Easier debugging.

File: src/darker/tests/test_multiline_strings.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+        (21, 23),
+        (23, 25),
+        (37, 39),
+        (39, 41),
+        (41, 43),
+        (43, 45),
+        (45, 47),
+        (47, 49),
+        (49, 55),
+    ]
```
Comment (MatthijsBurgh) [inline src/darker/tests/test_multiline_strings.py:87]: Maybe, provide a message, to show the values/diff. Easier debugging.

File: src/darker/tests/test_diff.py [hunk 3]
```diff
@@ -194,23 +195,64 @@ def test_opcodes_to_chunks():
...
+            [(start + 1, end + 1) for start, end in multiline_string_ranges],
+        )
+    )
+    # Normalize expected lines/ranges to 0-based, end-exclusive.
     expect_ranges = [[n, n] if isinstance(n, int) else n for n in expect]
     expect_linenums = list(chain(*(range(n[0], n[1] + 1) for n in expect_ranges)))
 
-    assert edit_linenums == expect_linenums
+    # Normalize result to 0-based, end-exclusive before comparison
+    assert [linenum - 1 for linenum in edit_linenums] == expect_linenums
```
Comment (MatthijsBurgh) [inline src/darker/tests/test_diff.py:250]: Maybe, provide a message, to show the values/diff. Easier debugging.

File: src/darker/tests/test_diff.py [hunk 4]
```diff
@@ -194,23 +195,64 @@ def test_opcodes_to_chunks():
...
 
 
 def test_opcodes_to_edit_linenums_empty_opcodes():
-    result = list(opcodes_to_edit_linenums([], context_lines=0))
+    result = list(
+        opcodes_to_edit_linenums([], context_lines=0, multiline_string_ranges=[])
+    )
 
     assert result == []
```
Comment (MatthijsBurgh) [inline src/darker/tests/test_diff.py:258]: Maybe, provide a message, to show the values/diff. Easier debugging.

File: src/darker/tests/test_git.py [hunk 5]
```diff
@@ -803,6 +804,51 @@ def test_edited_linenums_differ_revision_vs_lines(git_repo, context_lines, expec
     assert linenums == expect
```
Comment (MatthijsBurgh) [inline src/darker/tests/test_git.py:804]: Maybe, provide a message, to show the values/diff. Easier debugging.

File: src/darker/tests/test_git.py [hunk 6]
```diff
@@ -803,6 +804,51 @@ def test_edited_linenums_differ_revision_vs_lines(git_repo, context_lines, expec
...
+            "keep",
+            "CHANGED",
+        ]
+    )
+    revrange = git.RevisionRange("HEAD", ":WORKTREE:")
+    differ = git.EditedLinenumsDiffer(git_repo.root, revrange)
+
+    linenums = differ.revision_vs_lines(Path("a.py"), content, context_lines)
+
+    assert linenums == expect
```
Comment (MatthijsBurgh) [inline src/darker/tests/test_git.py:849]: Maybe, provide a message, to show the values/diff. Easier debugging.

File: src/darker/tests/test_main_reformat_single_file.py [hunk 7]
```diff
@@ -72,3 +72,51 @@ def test_reformat_single_file_common_ancestor(git_repo):
...
+        git_repo.root,
+        Path("a.py"),
+        RevisionRange("HEAD", ":WORKTREE:"),
+        rev2_content=TextDocument.from_str(modified),
+        rev2_isorted=TextDocument.from_str(modified),
+        enable_isort=False,
+        black_config={},
+    )
+
+    assert result.lines == tuple(expect.splitlines())
```
Comment (MatthijsBurgh) [inline src/darker/tests/test_main_reformat_single_file.py:122]: Maybe, provide a message, to show the values/diff. Easier debugging.

File: src/darker/tests/test_multiline_strings.py (hunk 1 above)
Comment (akaihola) [reply src/darker/tests/test_multiline_strings.py:124]: But doesn't Pytest already do this for us? If I e.g. change the expected value in the sixth case from `(2, 2)` to `(2, 3)`, Pytest gives me:
```python
$ pytest -k test_find_overlap
================ test session starts =================
platform linux -- Python 3.10.0, pytest-7.0.0, pluggy-1.0.0
rootdir: /home/kaiant/prg/darker, configfile: pytest.ini
plugins: mypy-0.9.0, isort-2.0.0, darker-0.1.2, kwparametrize-0.0.3, anyio-3.5.0, flake8-1.0.7
collected 713 items / 690 deselected / 23 selected                                                                                                                                                                                           

src/darker/tests/test_multiline_strings.py .....F.................                                                                                                                                                                     [100%]

====================== FAILURES ======================
__ test_find_overlap[1-3-ranges5-expect5] __

start = 1, end = 3, ranges = [(2, 2), (5, 6), (9, 11), (14, 17)], expect = (2, 3)

    @pytest.mark.kwparametrize(
        # `(start, end)` and `ranges` are end-exclusive
        # [...]
        dict(start=1, end=3, ranges=TEST_RANGES, expect=(2, 3)),
        # [...]
    )
    def test_find_overlap(start, end, ranges, expect):
        """`find_overlap()` finds the enclosing range for overlapping ranges"""
        result = multiline_strings.find_overlap(start, end, ranges)
    
>       assert result == expect
E       assert (2, 2) == (2, 3)
E         At index 1 diff: 2 != 3
E         Use -v to get the full diff

src/darker/tests/test_multiline_strings.py:124: AssertionError
============== short test summary info ===============
FAILED src/darker/tests/test_multiline_strings.py::test_find_overlap[1-3-ranges5-expect5] - assert (2, 2) == (2, 3)
=======================================
```

File: src/darker/tests/test_multiline_strings.py (hunk 2 above)
Comment (akaihola) [reply src/darker/tests/test_multiline_strings.py:87]: Here, too, for me Pytest already gives pretty verbose output if I change `(13, 15)` to `(12, 15)`:
```python
>       assert result == [
            # 1-based, end-exclusive
            (12, 15),
            (15, 17),
            (17, 19),
            (19, 21),
            (21, 23),
            (23, 25),
            (37, 39),
            (39, 41),
            (41, 43),
            (43, 45),
            (45, 47),
            (47, 49),
            (49, 55),
        ]
E       assert [(13, 15), (1...(23, 25), ...] == [(12, 15), (1...(23, 25), ...]
E         At index 0 diff: (13, 15) != (12, 15)
E         Use -v to get the full diff
```

And if I add the `-v` option, I get:
```python
E       AssertionError: assert [(13, 15), (1...(23, 25), ...] == [(12, 15), (1...(23, 25), ...]
E         At index 0 diff: (13, 15) != (12, 15)
E         Full diff:
E           [
E         -  (12, 15),
E         ?    ^
E         +  (13, 15),
E         ?    ^...
E         
E         ...Full output truncated (14 lines hidden), use '-vv' to show
```

...and with `-vv`:

```python
E       assert [(13, 15),\n (15, 17),\n (17, 19),\n (19, 21),\n (21, 23),\n (23, 25),\n (37, 39),\n (39, 41),\n (41, 43),\n (43, 45),\n (45, 47),\n (47, 49),\n (49, 55)] == [(12, 15),\n (15, 17),\n (17, 19),\n (19, 21),\n (21, 23),\n (23, 25),\n (37, 39),\n (39, 41),\n (41, 43),\n (43, 45),\n (45, 47),\n (47, 49),\n (49, 55)]
E         At index 0 diff: (13, 15) != (12, 15)
E         Full diff:
E           [
E         -  (12, 15),
E         ?    ^
E         +  (13, 15),
E         ?    ^
E            (15, 17),
E            (17, 19),
E            (19, 21),
E            (21, 23),
E            (23, 25),
E            (37, 39),
E            (39, 41),
E            (41, 43),
E            (43, 45),
E            (45, 47),
E            (47, 49),
E            (49, 55),
E           ]
```

**Comment by akaihola** [comment]
Moving to version [1.5.0](https://github.com/akaihola/darker/milestone/13) ([1.4.2](https://github.com/akaihola/darker/milestone/14) was just released).

File: src/darker/tests/test_multiline_strings.py (hunk 2 above)
Comment (MatthijsBurgh) [reply src/darker/tests/test_multiline_strings.py:87]: Check, I wasn't aware that assert in pytest gives a  good message

File: src/darker/tests/test_multiline_strings.py [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+            'f"double-quoted f-string"',
+            'r"double-quoted raw string"',
+            'rf"double-quoted raw f-string"',
+            'b"double-quoted bytestring"',
+            'rb"double-quoted raw bytestring"',  # (line 30)
+            '"""triple-double-quoted one-line string"""',
+            'f"""triple-double-quoted one-line f-string"""',
+            'r"""triple-double-quoted one-line raw string"""',
+            'rf"""triple-double-quoted one-line raw f-string"""',
+            'b"""triple-double-quoted one-line bytestring"""',  # (line 35)
```
[the remark this answers is not in the document]
Comment (akaihola) [reply src/darker/tests/test_multiline_strings.py:47]: The idea was to just have line number comments for the line ranges in the expected result. Then I noticed the long gaps before line 13 and on lines 25 to 36, and decided to add parenthesized running line numbers every five lines to ease visually scanning the document when debugging possible test failures.

If the parenthesized line numbers are confusing, I'd rather take them away (lines 1, 5, 10, 30, and 35).

Another possibility would be to have line number prefixes on every line and strip them out of the document before calling `get_multiline_string_ranges()`:

```python
    content = TextDocument.from_lines(
        [
            line.split(": ")[1]
            for line in [
                "1: 'single-quoted string'",
                "2: f'single-quoted f-string'",
                "3: r'single-quoted raw string'",
     ...
```

[the remark this answers is not in the document]
Comment (akaihola) [reply src/darker/tests/test_multiline_strings.py:47]: @MatthijsBurgh, or:
```python
    content = TextDocument.from_lines(
        [
            line for _line_number, line in [
                (1, "'single-quoted string'"),
                (2, "f'single-quoted f-string'"),
                (3, "r'single-quoted raw string'"),
     ...
```

Do you like one of these solutions better than the original?
